### PR TITLE
Fix #82, Empty path segments handled incorrectly

### LINKED ISSUE #82: Empty path segments handled incorrectly
```ocaml
utop # Uri.(to_string (resolve "" Uri.empty (of_string "/foo/bar/..")));;
- : bytes = "/foo/"
─( 13:45:36 )─< command 9 >──────────────────────────────────────────{ counter: 0 }─
utop # Uri.(to_string (resolve "" Uri.empty (of_string "/foo/bar//..")));;
- : bytes = "/foo/bar"
─( 13:45:40 )─< command 10 >─────────────────────────────────────────{ counter: 0 }─
utop # Uri.(to_string (resolve "" Uri.empty (of_string "/foo/bar///..")));;
- : bytes = "/foo/bar/"
─( 13:45:44 )─< command 11 >─────────────────────────────────────────{ counter: 0 }─
utop # Uri.(to_string (resolve "" Uri.empty (of_string "/foo/bar////..")));;
- : bytes = "/foo/bar//"
─( 13:45:48 )─< command 12 >─────────────────────────────────────────{ counter: 0 }─
utop # Uri.(to_string (resolve "" Uri.empty (of_string "/foo/bar//../baz")));;
- : bytes = "/foo/barbaz"
```

Suggested by @dbuenzli.

### PR DESCRIPTION
**Pull request by moonlightdrive**
Hi, hope this looks okay. Particularly the tests....

### CODE REVIEW

File: lib/uri.ml [hunk 1]
```diff
@@ -395,7 +395,8 @@ module Path = struct
       | [] when ascension > 0 -> List.rev_append
         ("/"::(rev_interject "/" Array.(to_list (make ascension "..")))) outp
       | [] -> List.(if length outp > 0 && hd outp = "/" then tl outp else outp)
-      | "/"::s::r when ascension > 0 -> loop (ascension - 1) outp r
+      | "/"::s::r when ascension > 0 ->
+        (if s = "/" then (s::r) else r) |> loop (ascension - 1) outp
```
Comment (dsheets) [inline lib/uri.ml]: We can't use `|>` due to legacy OCaml version support. If you could also put the `if` into a prior pattern like

```ocaml
| "/"::"/"::r when ascension > 0 -> loop (ascension - 1) outp ("/"::r)
| "/"::_::r when ascension > 0 -> loop (ascension - 1) outp r
```

that would be great.

File: lib_test/test_runner.ml [hunk 2]
```diff
@@ -269,6 +269,20 @@ let uri_rel_rel_res = [ (* relative-relative resolution *)
...
+  "/foo/bar/..", "/foo/";
+  "/foo/bar//..", "/foo/bar/";
+  "/foo/bar///..", "/foo/bar//";
+  "/foo/bar//../baz", "/foo/bar/baz"
+]
+
+let test_rel_empty_path_res =
+  List.map (fun (rel, res) ->
+      let test () = assert_equal ~printer:(fun l -> l)
+          res (Uri.to_string (Uri.resolve "" Uri.empty (Uri.of_string rel))) in
```
Comment (dsheets) [inline lib_test/test_runner.ml]: This can be written

```ocaml
Uri.(to_string (resolve "" empty (of_string rel)))
```

but it doesn't really matter.

**Comment by dsheets** [comment]
Thanks! I'm surprised at the size of the fix. I've left a couple line comments but otherwise it looks really good.

**Comment by moonlightdrive** [comment]
I made the changes you suggested.

What's the reasoning behind not using the if statement? Is it just more pleasant/legible without "if", or is it something else?

**Comment by avsm** [comment]
Thanks! Both the fix and the test cases look good to me.  There seems to be a strange failure in Yorick as regards use of `|>` though -- not sure what's going on there (or that it's related to this PR at all)

```
File "travis_opam.ml", line 146, characters 6-8:
Error: Unbound value |>
```

**Comment by dsheets** [comment]
This is due to ocaml/ocaml-travisci-skeleton#58 at <https://github.com/ocaml/ocaml-travisci-skeleton/pull/58/files#diff-9846f191f8d66f21dfc85443e92c56edR146>.

**Comment by dsheets** [comment]
There is no functional difference from removing the `if` -- it is a purely style issue to not nest conditionals inside of pattern matches. In particularly, it removes nesting (good) and makes it easier for compiler analysis to compile the pattern match (a sequence/tree of conditionals) efficiently.

**Comment by moonlightdrive** [comment]
Ah I see. Thank you.

**Pull request merged**